ansible/rancher/rancher-playbook.yml

```diff
@@ -0,0 +1,101 @@
+    terraform_workspace: "{{ lookup('env', 'TF_WORKSPACE') }}"
+    # Construct the path to the workspace's state file
+    terraform_state_file: "../../terraform/aws/cluster_nodes/terraform.tfstate.d/{{ terraform_workspace }}/terraform.tfstate"
+    fqdn: "{{ lookup('cloud.terraform.tf_output', 'fqdn', state_file=terraform_state_file) }}"
+  tasks:
+    - name: Check if kubeconfig file exists
+      stat:
+        path: "{{ kubeconfig_file }}"
+      register: kubeconfig_file_stat
+
+    - name: Fail if kubeconfig file does not exist
+      fail:
+        msg: "kubeconfig file '{{ kubeconfig_file }}' does not exist."
+      when: not kubeconfig_file_stat.stat.exists
+
+    - name: Install cert-manager CRDs
+      kubernetes.core.k8s:
+        kubeconfig: "{{ kubeconfig_file }}"
+        src: https://github.com/cert-manager/cert-manager/releases/download/v{{ cert_manager_version }}/cert-manager.crds.yaml
+        state: present
+      when: cert_manager_version != ""
+
+    - name: Install cert-manager
+      kubernetes.core.k8s:
+        kubeconfig: "{{ kubeconfig_file }}"
+        src: https://github.com/cert-manager/cert-manager/releases/download/v{{ cert_manager_version }}/cert-manager.yaml
+        state: present
+        wait: true
+        wait_timeout: 300
+      when: cert_manager_version != ""
+
+    - name: Wait for cert-manager to be ready
+      kubernetes.core.k8s_info:
+        kubeconfig: "{{ kubeconfig_file }}"
+        api_version: apps/v1
+        kind: Deployment
+        namespace: cert-manager
+        name: cert-manager
+      register: cert_manager_deployment
+      until: cert_manager_deployment.resources[0].status.readyReplicas == cert_manager_deployment.resources[0].status.replicas
+      retries: 10
+      delay: 30
+      when: cert_manager_version != ""
+
+    - name: Add Rancher Helm repository
+      kubernetes.core.helm_repository:
+        kubeconfig: "{{ kubeconfig_file }}"
+        name: "{{ rancher_chart_repo }}"
+        repo_url: "{{ rancher_chart_repo_url }}"
+      when: rancher_version != ""
+
+    - name: Install Rancher
+      kubernetes.core.helm:
+        kubeconfig: "{{ kubeconfig_file }}"
+        name: rancher
+        chart_ref: "{{ rancher_chart_repo }}/rancher"
+        release_namespace: cattle-system
+        create_namespace: true
+        chart_version: "{{ rancher_version }}"
+        wait: true
+        wait_timeout: "600s"
+        values:
+          hostname: "{{ fqdn }}"
+          bootstrapPassword: "{{ bootstrap_password }}"
+      when: rancher_version != ""
+
+    - name: Wait for Rancher to be ready
+      kubernetes.core.k8s_info:
+        kubeconfig: "{{ kubeconfig_file }}"
+        api_version: apps/v1
+        kind: Deployment
+        namespace: cattle-system
+        name: rancher
+      register: rancher_deployment
+      until: rancher_deployment.resources[0].status.readyReplicas == rancher_deployment.resources[0].status.replicas
```

Comment:
Is there a way to get rid of indexing for the resources? Off the top of my head, I want to say `kubectl` and `jq` or `yq`, but these are host-dependent, and hosts may not have them?

Reply:
No especially if we 're using the The _**kubernetes.core.k8s_info**_ module, this module returns a dictionary with a resources list to provide a consistent way to handle queries that might return one object or many. So in essence it's a safe call, if we can't index then the deployment failed anyway right? This a much cleaner way than installing jq, running a raw bash command to run kubectl and then parsing the output.